File: leetcode/Python/is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        color, visit = [0]*len(graph), [0]*len(graph)
        def bfs(v):
            if visit[v]>0:
                return True
            dq = collections.deque()
            dq.append(v)
            color[v] = visit[v] = 1
            while dq:
                cur = dq.popleft()
                val, nval = color[cur], 1
                if color[cur]==1:
                    nval += 1
                for ncur in graph[cur]:
                    if color[ncur]>0 and color[ncur]!=nval:
                        return False
                    elif visit[ncur]>0:
                        continue
                    dq.append(ncur)
                    color[ncur] = nval
                    visit[ncur] = 1
            return True
        for i in range(len(graph)):
            if not bfs(i):
                return False
        return True

File: leetcode/C++/is-graph-bipartite.cpp
class Solution {
public:
    bool bfs(int &v, vector<vector<int>>& graph, vector<int> &color, vector<int> &visit){
        if(visit[v]) return true;
        queue<int> q;
        int cur=0, val=0, nval=0;
        q.push(v); visit[v]=1; color[v]=1;
        while(!q.empty()){
            cur=q.front(); val=color[cur]; nval=(val==1)? 2 : 1; q.pop();
            for(int &ncur : graph[cur]){
                if(color[ncur]>0 && color[ncur]!=nval) return false;
                else if(visit[ncur]) continue;
                q.push(ncur); visit[ncur]=1; color[ncur]=nval;
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int sz = graph.size();
        vector<int> color(sz,0), visit(sz,0);
        for(int i=0; i<sz; ++i){
            if(!bfs(i, graph, color, visit)) return false;
        }
        return true;
    }
};